src/schedule/ll_schedule.c

```diff
@@ -285,7 +284,13 @@ static int schedule_ll_domain_set(struct ll_schedule_data *sch,
 		/* earlier periodic task, try to make it cadence-aligned with the existed task */
 		offset = (sch->domain->next_tick - task_start) %
 			 (sch->domain->ticks_per_ms * period / 1000);
-		task_start = task_start - task_start_ticks + offset;
+
+		/* start the period task not later than half period to avoid xrun */
+		deadline_delay_us = period >> 1;
+		if (offset > sch->domain->ticks_per_ms * deadline_delay_us)
+			offset -= sch->domain->ticks_per_ms * deadline_delay_us;
```

Comment:
Based on description I think the intention was to set offset to half period when it is later than half period: offset = sch->domain->ticks_per_ms * deadline_delay_us; not current offset minus half period.